File: src/net.common/token_bucket.cpp
#include "token_bucket.h"
#include "net.common/time.h"
#include <algorithm>
#include <stdexcept>

namespace net::common
{
    token_bucket::token_bucket(uint64_t capacity, double refill_interval_ms) :
        capacity(capacity),
        refill_interval_ms(refill_interval_ms)
    {
        state.store({ capacity, net::common::time::get_instance().timestamp() }, std::memory_order_relaxed);
    }

    bool token_bucket::consume(uint64_t amount)
    {
        token_state expected = state.load(std::memory_order_acquire);
        token_state desire;

        do
        {
            const uint64_t current_ms = net::common::time::get_instance().timestamp();
            const uint64_t elapsed_ms = current_ms - expected.last_ms;

            // 지난 시간만큼 token 추가
            const uint64_t current_tokens = std::min(capacity, expected.tokens);
            const uint64_t generated = static_cast<uint64_t>(elapsed_ms / refill_interval_ms);
            const uint64_t available_tokens = (generated >= capacity - current_tokens) ? capacity : current_tokens + generated;

            // 토큰 없다...
            // 클라 연결 거부
            if (available_tokens < amount)
            {
                return false;
            }

            // 토큰 있다...
            // 토큰 소모하고 정보 최신화
            desire.tokens = available_tokens - amount;
            desire.last_ms = (generated > 0) ? expected.last_ms + static_cast<uint64_t>(generated * refill_interval_ms) : expected.last_ms;
        } 
        while (!state.compare_exchange_weak(expected, desire, std::memory_order_release, std::memory_order_acquire));

        return true;
    }
}

File: Sources/TRPG.Common/TokenBucket.cs
using System;
using System.Threading;

namespace TRPG.Common;

/// <summary>
/// CAS 기반 token bucket rate limiter
/// </summary>
public sealed class TokenBucket
{
    /// <summary>
    /// 버킷 최대 사이즈
    /// </summary>
    private readonly ulong capacity;

    /// <summary>
    /// 토큰 1개가 회복되는 간격(ms)
    /// </summary>
    private readonly double refillIntervalMs;

    /// <summary>
    /// 현재 token 수 (Interlocked용)
    /// </summary>
    private long stateTokens;

    /// <summary>
    /// 마지막 token 갱신 시간(ms) (Interlocked용)
    /// </summary>
    private long stateLastMs;


    /// <summary>
    /// token bucket 생성
    /// </summary>
    public TokenBucket(ulong capacity, double refillIntervalMs)
    {
        this.capacity = capacity;
        this.refillIntervalMs = refillIntervalMs;

        stateTokens = (long)capacity;
        stateLastMs = (long)Time.Instance.Timestamp();
    }


    /// <summary>
    /// 지정한 token 수만큼 소모 시도
    /// </summary>
    public bool Consume(ulong amount = 1)
    {
        // CAS 루프
        while (true)
        {
            long currentTokens = Interlocked.Read(ref stateTokens);
            long currentLastMs = Interlocked.Read(ref stateLastMs);

            ulong currentMs = Time.Instance.Timestamp();
            ulong elapsedMs = currentMs - (ulong)currentLastMs;

            // 지난 시간만큼 token 추가
            ulong tokens = (ulong)Math.Min((long)capacity, currentTokens);
            ulong generated = (ulong)(elapsedMs / refillIntervalMs);
            ulong availableTokens = (generated >= capacity - tokens)
                ? capacity
                : tokens + generated;

            // 토큰 없다... 클라 연결 거부
            if (availableTokens < amount) return false;

            // 토큰 있다... 토큰 소모하고 정보 최신화
            long desiredTokens = (long)(availableTokens - amount);
            long desiredLastMs = (generated > 0)
                ? currentLastMs + (long)(generated * refillIntervalMs)
                : currentLastMs;

            if (Interlocked.CompareExchange(ref stateTokens, desiredTokens, currentTokens) == currentTokens)
            {
                Interlocked.CompareExchange(ref stateLastMs, desiredLastMs, currentLastMs);
                return true;
            }
            // CAS 실패 → 재시도
        }
    }
}
